File: typescript/lld/cache-design/rate-limiter/TokenBucketStrategy.ts
import { ClientState } from './ClientState';
import { RateLimitStrategy, RateLimitResult } from './RateLimitStrategy';

// Tokens refill continuously up to a cap; a request consumes one if
// available. Bursts are bounded by the cap, not by a hard window edge.
export class TokenBucketStrategy implements RateLimitStrategy {
    constructor(private capacity: number, private refillPerMs: number) {}

    initialState(): ClientState {
        return { tokens: this.capacity, lastRefill: Date.now() };
    }

    tryConsume(state: ClientState, now: number): RateLimitResult {
        const elapsed = Math.max(0, now - (state.lastRefill ?? now));   // clamp — clock can't go backward
        state.tokens = Math.min(this.capacity, (state.tokens ?? 0) + elapsed * this.refillPerMs);
        state.lastRefill = now;

        if (state.tokens < 1) {
            const retryAfterMs = Math.ceil((1 - state.tokens) / this.refillPerMs);
            return { allowed: false, retryAfterMs };
        }
        state.tokens -= 1;
        return { allowed: true };
    }
}

File: java/lld/cache-design/rate-limiter/TokenBucketStrategy.java
public class TokenBucketStrategy implements RateLimitStrategy {
    private final int capacity;
    private final double refillPerMs;

    public TokenBucketStrategy(int capacity, double refillPerMs) {
        this.capacity = capacity;
        this.refillPerMs = refillPerMs;
    }

    @Override
    public ClientState initialState() {
        ClientState s = new ClientState();
        s.tokens = capacity;
        s.lastRefill = System.currentTimeMillis();
        return s;
    }

    // Tokens refill continuously up to a cap; a request consumes one if
    // available. Bursts are bounded by the cap, not by a hard window edge.
    @Override
    public RateLimitResult tryConsume(ClientState state, long now) {
        long elapsed = Math.max(0, now - state.lastRefill);          // clamp — clock can't go backward
        state.tokens = Math.min(capacity, state.tokens + elapsed * refillPerMs);
        state.lastRefill = now;

        if (state.tokens < 1) {
            long retryAfterMs = (long) Math.ceil((1 - state.tokens) / refillPerMs);
            return new RateLimitResult(false, retryAfterMs);
        }
        state.tokens -= 1;
        return new RateLimitResult(true, 0);
    }
}